src/hb-ot-shape-complex-arabic-fallback.hh

```diff
@@ -254,6 +256,137 @@ struct arabic_fallback_plan_t
   OT::hb_ot_layout_lookup_accelerator_t accel_array[ARABIC_FALLBACK_MAX_LOOKUPS];
 };
 
+static OT::SubstLookup *
+arabic_fallback_pua1_synthesize_lookup_single (const hb_ot_shape_plan_t *plan HB_UNUSED,
+					       hb_font_t *font,
+					       unsigned int feature_index)
+{
+  OT::HBGlyphID glyphs[SHAPING_TABLE_PUA1_LEN];
+  OT::HBGlyphID substitutes[SHAPING_TABLE_PUA1_LEN];
+  /* Each glyph takes four bytes max, and there's some overhead. */
+  char buf[SHAPING_TABLE_PUA1_LEN * 4 + 128];
+
+  return synthesize_lookup_single (plan,
+				   font,
+				   feature_index,
+				   shaping_table_pua1,
+				   SHAPING_TABLE_PUA1_FIRST,
+				   SHAPING_TABLE_PUA1_LAST,
+				   glyphs,
+				   substitutes,
+				   buf,
+				   sizeof (buf));
+}
+
+static OT::SubstLookup *
+arabic_fallback_pua2_synthesize_lookup_single (const hb_ot_shape_plan_t *plan HB_UNUSED,
+					       hb_font_t *font,
+					       unsigned int feature_index)
+{
+  OT::HBGlyphID glyphs[SHAPING_TABLE_PUA2_LEN];
+  OT::HBGlyphID substitutes[SHAPING_TABLE_PUA2_LEN];
+  /* Each glyph takes four bytes max, and there's some overhead. */
+  char buf[SHAPING_TABLE_PUA2_LEN * 4 + 128];
+
+  return synthesize_lookup_single (plan,
+				   font,
+				   feature_index,
+				   shaping_table_pua2,
+				   SHAPING_TABLE_PUA2_FIRST,
+				   SHAPING_TABLE_PUA2_LAST,
+				   glyphs,
+				   substitutes,
+				   buf,
+				   sizeof (buf));
+}
+
+static OT::SubstLookup *
+arabic_fallback_pua1_synthesize_lookup_ligature (const hb_ot_shape_plan_t *plan HB_UNUSED,
+						 hb_font_t *font)
+{
+  OT::HBGlyphID first_glyphs[ARRAY_LENGTH_CONST (ligature_table_pua1)];
+  unsigned int first_glyphs_indirection[ARRAY_LENGTH_CONST (ligature_table_pua1)];
+  unsigned int ligature_per_first_glyph_count_list[ARRAY_LENGTH_CONST (first_glyphs)];
+
+  /* We know that all our ligatures are 2-component */
+  OT::HBGlyphID ligature_list[ARRAY_LENGTH_CONST (first_glyphs) * ARRAY_LENGTH_CONST(ligature_table_pua1[0].ligatures)];
+  unsigned int component_count_list[ARRAY_LENGTH_CONST (ligature_list)];
+  OT::HBGlyphID component_list[ARRAY_LENGTH_CONST (ligature_list) * 1/* One extra component per ligature */];
+
+  /* 16 bytes per ligature ought to be enough... */
+  char buf[ARRAY_LENGTH_CONST (ligature_list) * 16 + 128];
+
+  return synthesize_lookup_ligature (plan,
+				     font,
+				     first_glyphs,
+				     first_glyphs_indirection,
+				     ligature_per_first_glyph_count_list,
+				     ligature_list,
+				     component_count_list,
+				     component_list,
+				     ligature_table_pua1,
+				     ARRAY_LENGTH_CONST (ligature_table_pua1),
+				     buf,
+				     sizeof (buf),
+				     0);
```

Comment:
This 0 is for lookup flags, there are a few mark ligatures so can’t use `IgnoreMarks`. I should have probably split the marks into there own looks, but by testing shows that Uniscribe breaks the ligatures when there are marks, so the bad behaviour is actually closer to what it does.